clifford/invariant_decomposition.py

```diff
@@ -0,0 +1,184 @@
+"""
+.. currentmodule:: clifford.invariant_decomposition
+
+=====================================================
+invariant_decomposition (:mod:`clifford.invariant_decomposition`)
+=====================================================
+
+.. versionadded:: 1.4.0
+
+
+This file implements the invariant decomposition (aka bivector split) of bivectors into
+mutually commuting orthogonal simple bivectors, based on the method of :cite:`roelfs2021thesis`, chapter 6.
+
+The invariant decomposition enables closed form exponentials and logarithms, and the factorization of
+rotors into simple rotors.
+
+Example usage::
+
+>>> B = 1*e12 + 2*e34
+>>> bivector_split(B)
+[1^e12, 2^e34]
+
+Implemented functions
+---------------------
+
+.. autofunction:: bivector_split
+.. autofunction:: rotor_split
+.. autofunction:: exp
+.. autofunction:: log
+
+
+Helper functions
+----------------
+
+.. autofunction:: _bivector_split
+.. autofunction:: single_split
+
+"""
+import math
+from functools import reduce
+
+import numpy as np
+
+from ._settings import _eps
+from . import _numba_utils
+
+
+@_numba_utils.njit(cache=True)
+def _single_split_even_values(Wm_array, li, r):
+    ND = np.zeros((2, Wm_array[0, :].shape[0]), dtype=np.complex_)
+    for i in range(0, Wm_array.shape[0]//2+1):
+        ND[0, :] += Wm_array[2*i, :] * li**(r - i)
+    for i in range(0, Wm_array.shape[0]//2):
+        ND[1, :] += Wm_array[2*i+1, :] * li**(r - i - 1)
+    return ND
+
+
+@_numba_utils.njit(cache=True)
+def _single_split_odd_values(Wm_array, li, r):
+    ND = np.zeros((2, Wm_array[0, :].shape[0]), dtype=np.complex_)
+    for i in range(0, Wm_array.shape[0]//2):
+        ND[0, :] += Wm_array[2 * i + 1, :] * li ** (r - i)
+        ND[1, :] += Wm_array[2*i, :] * li**(r - i)
+    return ND
+
+
+def single_split_even(Wm, li, r):
+    """Helper function to compute a single split for a given set of W_m and
+    eigenvalue lambda_i, when the total number of terms in the split is even.
+    """
+    Wm_array = np.array([W.value for W in Wm])
+    ND = _single_split_even_values(Wm_array, li, r)
+    N = Wm[0].layout.MultiVector(ND[0, :])
+    D = Wm[0].layout.MultiVector(ND[1, :])
+    return N*D.leftLaInv()
+
+
+def single_split_odd(Wm, li, r):
+    """Helper function to compute a single split for a given set of W_m and
+    eigenvalue lambda_i, when the total number of terms in the split is odd.
+    """
+    Wm_array = np.array([W.value for W in Wm])
+    ND = _single_split_odd_values(Wm_array, li, r)
+    N = Wm[0].layout.MultiVector(ND[0, :])
+    D = Wm[0].layout.MultiVector(ND[1, :])
+    return N*D.leftLaInv()
```

Comment:
@hugohadfield, I assume this jitted code is your doing. Is there a reason you define the `values` function vs just using the multivector jit support?

Reply:
Maybe #412 will resolve this.